[Site Admin]: Add Link component with stateful navigation (ARC-24)

retrofox opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

This PR adds a `<Link />` component built on top of the internal `useLink()` hook to handle client-side navigation while preserving the navigation state. It simplifies usage and avoids repeating logic in multiple places.

Two stories were added to the Storybook:
* A basic usage example.
* One that shows how to pass a custom state


Related to #

## Proposed Changes

* [Site Admin]: Add Link component with stateful navigation

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

*

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Run the Site Admin Storybook application
```sh
yarn workspace @automattic/site-admin run storybook
```
2. Go to the Link stories
3. Use the Pass Custom State story to confirm how the state is preserved during navigation.

![site-admin-link-component-2](https://github.com/user-attachments/assets/f88362e9-c6f8-444a-8696-6526b280cb0b)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

retrofox committed:
Fix RouterProvider nesting in README examples

retrofox committed:
Add Link component for router navigation
Introduce new Link component with useLink hook integration and basic story examples

retrofox committed:
wrap the story with a RouterProvider instance
It's required to make the navigation system works.

retrofox committed:
add a new PassCustomState Link story

retrofox committed:
tweak colors to Link stories

retrofox committed:
Improve Link component type definitions
Add explicit LinkProps type and clean up component props interface for better TypeScript support

retrofox committed:
Add documentation and improve styles for Link

psealock commented on packages/site-admin/src/router/components/link/index.tsx:
```diff
@@ -0,0 +1,21 @@
+			{ children }
+		</a>
+	);
+}
```
The idea is to expose the `Link` component for use right? In this case, you'll need to export it in `packages/site-admin/src/index.ts`. And if thats the case, also a description in the README would be good.

retrofox replied:
That's correct. Let's expose it.

retrofox replied:
https://github.com/Automattic/wp-calypso/pull/101820/commits/2d71c238986b70fc3e4c89ee273bf647e26c3340

Added also in the Changelog file. I'm going to create an issue to document the API of the package, too.

retrofox commented:
<!-- linear:isThreadRoot -->

That's correct. Let's expose it.

retrofox committed:
Export Link component from site-admin package root

mirka commented on packages/site-admin/CHANGELOG.md:
```diff
@@ -2,25 +2,27 @@
 
+### Components
 - `SiteHub`: Site navigation and context switcher
+- `Link`: Router component providing declarative navigation
```
We don't have one now, but I'm thinking we may want to have some kind of `Link` component in broader Calypso. Could we maybe rename this to `SiteLink` or something, to avoid potential naming collisions?

retrofox replied:
If we stick to the idea that Calypso is just used as a place to hold the code, I think it’s fine to keep the Link component name. In case of name collisions, can we always use an alias?<br><br>otherwise, I’d go with something like `<NavLink />` or something that clearly refers to the navigation link concept.

mirka replied:
Not specifically a Calypso problem, because `Link` components are very common in frameworks/apps. I guess my question is, will this site-admin (and thus the site-admin `Link`) never be a mere subsection of a broader app, where let's say you want to put a Next.js `Link` and a site-admin `Link` in the same JS file? (The answer could be "no", just asking 😄)

retrofox replied:
Fair question.
I’m not sure what name we could use instead, though.  `NavLink,` for instance, feels confusing to me since it sounds like it's tied to a navigation section. Could `RouterLink` work?

mirka replied:
If it's a "sub-router" specifically for site-admin, then maybe it should be called `SiteAdminLink`? And if it's a more general-use router, should it live somewhere else where it can be called something more generic? Or is there already an existing router we can reuse here?

Non-blocking, just some architecture thoughts.

mirka commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,110 @@
+	parameters: {
+		docs: {
+			description: {
+				story:
+					'The Link component enables navigation between routes in the application, maintaining state and avoiding full page reloads.',
```
Tip: You can write these descriptions as a JSDoc, and it works the same!

```js
/**
 * The description.
 */
export const Default = {
```

retrofox replied:
Nice one. Updating…

retrofox replied:
Actually it doesn't work in my local when running storybook app, Lena :-/

mirka replied:
It's already an [official feature](https://storybook.js.org/docs/7/api/doc-blocks/doc-block-description#writing-descriptions) in v7 🥺 

<img width="1041" alt="Story description" src="https://github.com/user-attachments/assets/78e0cbdc-403e-4644-b409-3245baecfea0" />


```diff
diff --git a/packages/site-admin/src/router/components/link/stories/index.stories.tsx b/packages/site-admin/src/router/components/link/stories/index.stories.tsx
index c01beabfbb2..0fa5ed833cb 100644
--- a/packages/site-admin/src/router/components/link/stories/index.stories.tsx
+++ b/packages/site-admin/src/router/components/link/stories/index.stories.tsx
@@ -29,20 +29,17 @@ export default meta;
 
 type Story = StoryObj< typeof Link >;
 
+/**
+ * The Link component enables navigation between routes in the application, maintaining state and avoiding full page reloads.
+ *
+ * Hello world!
+ */
 export const Default: Story = {
 	args: {
 		to: '/home',
 		children: 'Homepage',
 		className: 'story-link',
 	},
-	parameters: {
-		docs: {
-			description: {
-				story:
-					'The Link component enables navigation between routes in the application, maintaining state and avoiding full page reloads.',
-			},
-		},
-	},
 };
 
 export const PassCustomState: Story = {
```

retrofox committed:
Improve Link story with better state management

retrofox committed:
Update Link story documentation format

ciampo commented on packages/site-admin/CHANGELOG.md:
```diff
@@ -2,25 +2,27 @@
 - `SiteIcon`: Site icon display component
 
-## Routing System
+### Routing System
```
Tiny nit: let's make sure we always add an empty line after headings

ciampo commented on packages/site-admin/CHANGELOG.md:
```diff
@@ -2,25 +2,27 @@
 - `useLocation`: Hook for accessing current route location
 - `useMatch`: Hook for route matching functionality
 
 ## Next
```
Not related to this PR specifically, but I'm not sure I get the sense for the "Next" section?

I guess the README should always reflect a snapshot of the codebase, and we could otherwise have a CHANGELOG (or other ways) to track releases and what they included?

retrofox replied:
To clarify, this is the CHANGELOG.md file, not the README.md

ciampo replied:
🤦  Got it. I got mixed up.

ciampo commented on packages/site-admin/src/router/components/link/index.tsx:
```diff
@@ -0,0 +1,21 @@
+	children: React.ReactNode;
+};
+
+export function Link( { to, options, children, ...props }: LinkProps ) {
```
We should either omit `onClick` from the accepted props, or make sure that we call `props?.onClick( e )` on the anchor tag. 

Currently, if a consumer passes the `onClick` prop, it will be ignored

retrofox replied:
I think if the consumer passes the onClick prop, it won’t be ignored since it will override the onClick callback provided by the useLink hook: `<a href={ href } onClick={ onClick } { ...props }>`.

ciampo replied:
I may not explained myself correctly — the `onClick` callback returned by `useLink` will be ignored, in that case. I'm not sure that's the behaviour that we expect?

retrofox replied:
gotcha. I suggest removing the onClick and href props from the component and relying solely on the props returned by the useLink() hook.

mirka replied:
Don't we want to merge the `onClick` props? I'm imagining some people will want to attach click listeners for analytics.

retrofox replied:
That's a good point. It could be outside the scope of this component.
Because you could create a new one using the useLink() hook.
For the tracking issue, should we try to do it at the context level?  By using the `beforeNavigate` property.

mirka replied:
Is there a reason we _don't_ want to merge the `onClick`s here? I'd say it's a pretty common pattern — we should simply do it if there's no reason not to. Not tested, but literally just something like:

```jsx
onClick={
  ( event ) => {
    props.onClick( event );
    onClick( event );
  }
}
```

retrofox replied:
I don't have any particular objection.

retrofox replied:
I wonder whether we should create a handler function wrapped with a useCallback() instance.

retrofox replied:
Added onClick support. Also, I've added a story that shows how to use the Link component with a custom onClick callback function.

retrofox replied:
<img width="982" alt="image" src="https://github.com/user-attachments/assets/8f27a7cb-da0e-456e-b4cb-756e7bfe163f" />

mirka replied:
> I wonder whether we should create a handler function wrapped with a useCallback() instance.

I'd say the standard "best practice" is to not do it preemptively, especially in this kind of use case where `Link` children is unlikely to be doing something costly to re-render, since `useCallback` itself is not free. Though, of course the unnecessary `useCallback`s probably won't be a perceivable net negative on performance up until a certain number.

retrofox replied:
Yeah… I have to admit I wasn't sure whether to add it or not 😕<br>@<!-- linear:userMention:dd042a66-b175-4094-a734-0b7b31035597 -->riad do you have an opinion here?

ciampo replied:
There may be a glitch in how Linear connects usernames — Riad is not tagged correctly on GitHub.

My thoughts:
- We should formward/merge the `onClick` prop — which we now do ✅ 
- Wrapping the handler in `useCallback` fees like premature optimization, so I'd personally wouldn't add it. Having said that, the difference is likely negligible. I'll leave it up to @retrofox , and in any case, it's not a big deal and we shouldn't worry too much about it

retrofox replied:
Thanks, both. Let's remove the hook for now.

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+	decorators: [
+		function WithRouterProvider( Story ) {
+			return (
+				<RouterProvider routes={ [] } pathArg="p">
```
Why is `routes` an empty array? Can we change this example to be more similar to real-world scenario?

retrofox replied:
It's true it feels a bit wired. Anyway, having an empty routes array can be a valid example for a story?<br>We could also make the routes prop optional and default it to an empty array, too. We need to provider to pass the context  to the story.

ciampo replied:
I'm sure it is possible to leave an empty array, although I believe that we should try to put together the most comprehensive (and close to standard use cases) examples for the consumers of the package. And in that sense, I imagine that it would be better to have a list of routes and navigate between them?

retrofox replied:
Putting it all together will make the story more complicated to understand. Also, although clearly, the routes are a fundamental piece of the implementation, they are not part of the Link component.

The benefit of this component, and what we should highlight, is the usage of its options (NavigationOptions). The second story covers it.

I would suggest that we improve other stories to show the whole implementation in detail. We already have an issue for this.

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+	decorators: [
+		function WithRouterProvider( Story ) {
+			return (
+				<RouterProvider routes={ [] } pathArg="p">
```
What's the `pathArg` prop for? Is it necessary? Should we at least document it / add a comment?

retrofox replied:
Let me make a quick update/iteration on it

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+type Story = StoryObj< typeof Link >;
+
+/**
+ * The Link component enables navigation between routes in the application,<br />
```
Do we need `<br />`s in the JS comments?

retrofox replied:
yes 🙁

mirka replied:
Why...? If we opted to hardcode line breaks in all Markdown readmes and JSDocs, it would be a mess. This is not poetry. Line wrapping is for the presentation layer to handle 😄

retrofox replied:
The thing is if I don't add the `<br />,` it gets rendered in a single line, which is confusing:

<img src="https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/7859b8c4-3b4f-4e70-bcc0-f19748e2baeb/e156b23b-0f82-4c7b-b397-ad7152577c9e?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi83ODU5YjhjNC0zYjRmLTRlNzAtYmNjMC1mMTk3NDhlMmJhZWIvZTE1NmIyM2ItMGY4Mi00YzdiLWIzOTctYWQ3MTUyNTc3YzllIiwiaWF0IjoxNzQzNDE5MTIwLCJleHAiOjMzMzEzOTc5MTIwfQ.Q2r66ymfzrRGWZZAFpLzDz7gOO7e7k363rkg9lJaEsY " alt="image.png" width="895" data-linear-height="143" />

If I add break lines, it adds two lines per row:

<img src="https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/02ffc24c-972e-49cc-bd8e-c4436289bc60/e5bc3a8d-db55-4ace-ac46-fe702905be63?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi8wMmZmYzI0Yy05NzJlLTQ5Y2MtYmQ4ZS1jNDQzNjI4OWJjNjAvZTViYzNhOGQtZGI1NS00YWNlLWFjNDYtZmU3MDI5MDViZTYzIiwiaWF0IjoxNzQzNDE5MTIwLCJleHAiOjMzMzEzOTc5MTIwfQ.OzlhuNGw3lq3TiUdxtAmL4uB_iurkcJdfvXpDf7kp2s " alt="image.png" width="631" data-linear-height="206" />

retrofox replied:
removed `<br />` s

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+		useLocation();
+
+		// Pick the location state from the browser history state
+		const { init, history } = ( browserHistory.location.state || {
+			luckyNumber: 0,
+			init: true,
+			history: [],
+		} ) as unknown as {
+			luckyNumber: number;
+			init: boolean;
+			history: number[];
+		};
```
Should we extract this to a variable outside of the render function, something like

```
const INITIAL_HISTORY_STATE = {
	luckyNumber: 0,
	init: true,
	history: [],
}
```

We should also extract the type, so that it can be used across the file if needed.

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+			luckyNumber: 0,
+			init: true,
+			history: [],
+		} ) as unknown as {
```
I'd like to think about ways to improve the DevX when working with link / routers / history:

- having to type cast `history.location` is not great;
- having to manually implement a `init` and `history.action` check is also quite elaborate for a potential user
- manually handling the history array also feels somewhat hacky

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+		/*
+		 * Create a new lucky number only when the user clicks on the link.
+		 */
+		if ( ! init && browserHistory.action !== 'POP' ) {
+			newLuckyNumber = ( ( Math.random() * 100 ) | 0 ) + 1;
+		}
```
Is it possible that any re-render of this component is going to create a new lucky number, instead of only when the link is clicked?

retrofox committed:
Add line breaks between sections in the CHANGELOG

retrofox committed:
fix comment about creating new lucky number

retrofox committed:
Add comment explaining useLocation in Link story

retrofox committed:
remove href and onClick from Link component

retrofox committed:
Make RouterProvider props optional with defaults
Add default values for routes and pathArg props with improved JSDoc documentation

retrofox committed:
Add type definitions for Link story state management
Create HistoryState type and initial state constant for better code organization

retrofox committed:
Simplify Link story state management
Remove init flag and rename history array to prevLuckyNumbers for better clarity

retrofox committed:
use onClick prop to get a lucky number

retrofox committed:
add a WithCustomOnClick story

retrofox committed:
add support for custom onClick prop

retrofox committed:
add style to Link story

retrofox committed:
organize next Changelog section

retrofox commented:
@mirka @ciampo I'd like to merge this PR as soon as possible. We've addressed nearly all the feedback. Let us know whether there is something that could be consider like a blocker.

Regarding the stories: we're planning to add a Documentation App store that will demonstrate how to use the package resources. wip https://github.com/Automattic/wp-calypso/pull/102193

mirka commented:
No blockers from me 👍

retrofox committed:
Simplify Link component click handler
Remove useCallback and simplify event handling

ciampo commented on packages/site-admin/src/router/components/link/stories/index.stories.tsx:
```diff
@@ -0,0 +1,131 @@
+					{ ...args }
+					options={ { state } }
+					className="story-link"
+					onClick={ () => setNewLuckyNumber( ( ( Math.random() * 10 ) | 0 ) + 1 ) }
```
We should call `args.onClick`, otherwise it will be never logged in the Storybook actions

retrofox commented:
> more in general, I would expect a more canonical example, where there are multiple links that cause the router state to change

It will increase the complexity of the story. I prefer to keep the story simple and work on a more complex story https://github.com/Automattic/wp-calypso/pull/102193

ciampo commented:
Agree that we need to keep stories as simple as possible. What I'm trying to say, is that I'm not sure how much a consumer can learn from an example story where the `Link` is basically used as a button generating a random number. It doesn't really show how to use `Link` as a "Router component providing declarative navigation"

retrofox commented:
<!-- linear:replyTo:2750316535 -->

it shows how to propagate the state though the Link props. We can remove the simplets story

ciampo commented:
<!-- linear:replyTo:2750316535 -->

> The third one, something you asked about, but it’s already there, shows how to pass a custom callback for the onClick event.

I didn't ask about a specific story — in fact, we should drop it in favour of using Storybook actions (see [ARC-41](https://linear.app/a8c/issue/ARC-41/fix-forwarding-of-onclick-in-the-story#comment-ae633ddc))
